File: M/1031-maximum-sum-of-two-non-overlapping-subarrays/main.go
package main

import "fmt"

func maxSumTwoNoOverlap(nums []int, firstLen int, secondLen int) int {
	buf := make([]int, len(nums))
	sum := 0
	for i := 0; i < len(nums); i++ {
		sum += nums[i]
		buf[i] = sum
	}
	left1 := make([]int, len(nums))
	max := 0
	for i := firstLen - 1; i < len(nums); i++ {
		v := 0
		if i >= firstLen {
			v = buf[i] - buf[i-firstLen]
		} else {
			v = buf[i]
		}
		if v > max {
			max = v
		}
		left1[i] = max
	}
	right1 := make([]int, len(nums))
	max = 0
	for i := len(nums) - secondLen; i >= 0; i-- {
		v := 0
		if i > 0 {
			v = buf[i+secondLen-1] - buf[i-1]
		} else {
			v = buf[i+secondLen-1]
		}
		if v > max {
			max = v
		}
		right1[i] = max
	}
	
	ret := 0
	for i := 0; i < len(nums)-1; i++ {
		if ret < left1[i]+right1[i+1] {
			ret = left1[i] + right1[i+1]
		}
	}

	left2 := make([]int, len(nums))
	max = 0
	for i := secondLen - 1; i < len(nums); i++ {
		v := 0
		if i >= secondLen {
			v = buf[i] - buf[i-secondLen]
		} else {
			v = buf[i]
		}
		if v > max {
			max = v
		}
		left2[i] = max
	}
	right2 := make([]int, len(nums))
	max = 0
	for i := len(nums) - firstLen; i >= 0; i-- {
		v := 0
		if i > 0 {
			v = buf[i+firstLen-1] - buf[i-1]
		} else {
			v = buf[i+firstLen-1]
		}
		if v > max {
			max = v
		}
		right2[i] = max
	}
	for i := 0; i < len(nums)-1; i++ {
		if ret < left2[i]+right2[i+1] {
			ret = left2[i] + right2[i+1]
		}
	}
	return ret
}

func main() {
	fmt.Println()
}

File: M/1031-maximum-sum-of-two-non-overlapping-subarrays/main.rs
struct Solution {}

impl Solution {
  pub fn max_sum_two_no_overlap(nums: Vec<i32>, first_len: i32, second_len: i32) -> i32 {
    let mut left: Vec<i32> = vec![0; nums.len()];
    let mut right: Vec<i32> = vec![0; nums.len()];

    let mut sum: i32 = 0;
    let mut r_sum: i32 = 0;
    (0..first_len).for_each(|idx| {
      sum += nums[idx as usize];
      r_sum += nums[nums.len() - 1 - idx as usize];
    });

    let mut max_sum = sum;
    let mut max_r_sum = r_sum;
    (first_len as usize..=nums.len() - second_len as usize).for_each(|idx| {
      left[idx] = max_sum;
      sum -= nums[idx - first_len as usize];
      sum += nums[idx];
      if max_sum < sum {
        max_sum = sum;
      }

      right[nums.len() - idx] = max_r_sum;
      r_sum -= nums[nums.len() - 1 - idx + first_len as usize];
      r_sum += nums[nums.len() - 1 - idx];
      if max_r_sum < r_sum {
        max_r_sum = r_sum;
      }
    });

    let mut ret: i32 = 0;
    let mut second_sum: i32 = 0;
    (0..second_len as usize - 1).for_each(|idx| {
      second_sum += nums[idx];
    });
    (second_len as usize - 1..nums.len()).for_each(|idx| {
      second_sum += nums[idx];
      if idx >= second_len as usize {
        second_sum -= nums[idx - second_len as usize];
      }

      if idx >= (first_len + second_len - 1) as usize
        && second_sum + left[idx - second_len as usize + 1] > ret
      {
        ret = second_sum + left[idx - second_len as usize + 1];
      }
      if idx + (first_len as usize) < nums.len() && second_sum + right[idx + 1] > ret {
        ret = second_sum + right[idx + 1];
      }
    });

    ret
  }
}

fn main() {
  println!(
    "{}",
    Solution::max_sum_two_no_overlap(vec![0, 6, 5, 2, 2, 5, 1, 9, 4], 1, 2)
  );
  println!(
    "{}",
    Solution::max_sum_two_no_overlap(vec![3, 8, 1, 3, 2, 1, 8, 9, 0], 3, 2)
  );
  println!(
    "{}",
    Solution::max_sum_two_no_overlap(vec![2, 1, 5, 6, 0, 9, 5, 0, 3, 8], 4, 3)
  );
}
